MINOR: Adds KRaft versions of most streams system tests

AlanConfluent opened the pull request:
Adds the annotation `@matrix(metadata_quorum=quorum.all_non_upgrade)` to many existing tests, which runs them with each of zookeeper and remote_kraft nodes.

This skips tests which use various forms of Kafka versioning since those seem to have issues with KRaft at the moment.  Running these tests with KRaft will require a followup PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

AlanConfluent committed:
MINOR: Adds KRaft versions of most streams system tests

vvcephei commented on tests/kafkatest/tests/streams/streams_broker_bounce_test.py:
```diff
@@ -205,11 +211,17 @@ def collect_results(self, sleep_time_secs):
     @matrix(failure_mode=["clean_shutdown", "hard_shutdown", "clean_bounce", "hard_bounce"],
             broker_type=["leader", "controller"],
             num_threads=[1, 3],
             sleep_time_secs=[120])
```
What do you think about explicitly specifying `quorum.zk` instead of relying on the default? It's maybe mildly more verbose, but it avoids having to think through overrides when you're debugging a test failure out of the blue.

If you're in favor, I'd suggest just removing all the default parameter values that got added here to be sure everything is explicitly configured.

AlanConfluent replied:
I added this extra matrix configuration to avoid `broker_type=="controller"` for `metadata_quorum=="quorum.remote_kraft"`-- not sure if you noticed that, otherwise I would have just added `metadata_quorum=quorum.all_non_upgrade` to the existing matrix to cover both modes as you said.  This is because killing the controller when it's separate from the broker is maybe not what the test is attempting to test and also requires other changes to accommodate killing off the separate KRaft controller node, but that seems a bit overkill for this purpose.

Is that what

vvcephei replied:
Thanks, @AlanConfluent . I was actually making a slightly different suggestion, namely to explicitly include `metadata_quorum=[quorum.zk]` in this second matrix config, and to remove the default parameter value `metadata_quorum=quorum.zk` from the method signature.

Github charmingly won't let me do a suggestion to make this clearer because the diff includes a deleted line.

AlanConfluent replied:
I see.  Done!

vvcephei commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -73,9 +77,9 @@ def test_standby_tasks_rebalance(self):
-        self.wait_for_verification(processor_1, "ACTIVE_TASKS:2 STANDBY_TASKS:[1-3]", processor_1.STDOUT_FILE)
-        self.wait_for_verification(processor_2, "ACTIVE_TASKS:2 STANDBY_TASKS:[1-3]", processor_2.STDOUT_FILE)
-        self.wait_for_verification(processor_3, "ACTIVE_TASKS:2 STANDBY_TASKS:[1-3]", processor_3.STDOUT_FILE)
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_1.STDOUT_FILE)
+        self.wait_for_verification(processor_2, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_2.STDOUT_FILE)
+        self.wait_for_verification(processor_3, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_3.STDOUT_FILE)
```
What's up with this verification change? It looks like you're increasing the specificity of the verification here, which is good. I'm wondering if it has some relationship to the quorum change, of if it's just on the side.

AlanConfluent replied:
Ahh, I know how this change got made.  I had made my changes in another repo locally and copied the file over to this one...  Must have been made by another commit in the original repo.  Will revert this and make sure there are no similar changes elsewhere.

vvcephei reviewed (changes requested):
Thanks, @AlanConfluent !

It's great to see us tightening the knobs on KRaft.

In addition to addressing the review comments, can you post a link to the system test results with this change? It would be good to verify the impact before merging.

Thanks!
-John

AlanConfluent committed:
Remove mistaken copy

AlanConfluent committed:
Remove extra node added

AlanConfluent committed:
Fixes test to avoid waiting on ZK with KRaft

guozhangwang commented:
> In addition to addressing the review comments, can you post a link to the system test results with this change? It would be good to verify the impact before merging.

+1. I'd also love to learn how much system test time increase this one would incur.

guozhangwang commented on tests/kafkatest/tests/streams/streams_broker_bounce_test.py:
```diff
@@ -251,8 +264,9 @@ def test_broker_type_bounce_at_start(self, failure_mode, broker_type, sleep_time
-    def test_many_brokers_bounce(self, failure_mode, num_failures):
+            num_failures=[2],
+            metadata_quorum=quorum.all_non_upgrade)
+    def test_many_brokers_bounce(self, failure_mode, num_failures, metadata_quorum=quorum.zk):
```
nit: since we already set the value range of `metadata_quorum` in the matrix, do we still need to set its default as `quorum.zk`? Seems the default value would never be used? Ditto elsewhere.

guozhangwang replied:
Ah just saw @vvcephei 's comment earlier, that makes sense. Please feel free to ignore.

AlanConfluent replied:
Yeah, I think if it's not defined, it'll default to using zk, but I just wanted to be a bit more explicit about the default.

guozhangwang commented on tests/kafkatest/tests/streams/streams_broker_bounce_test.py:
```diff
@@ -205,11 +212,17 @@ def collect_results(self, sleep_time_secs):
+            broker_type=["leader"],
+            num_threads=[1, 3],
+            sleep_time_secs=[120],
+            metadata_quorum=[quorum.remote_kraft])
```
Why we only want to test remote_kraft but not collocated kraft?

AlanConfluent replied:
That's what was suggested for tests which are not testing core Kafka functionality and could be affected, which I don't think this is.  For those, the predefined variable `quorum.all_non_upgrade` exists and is equal to `[quorum.remote_kraft, quorum.zk]`.  Also, given how long these tests run, I didn't want to run more variants than is necessary.

guozhangwang replied:
Cool, thanks!

guozhangwang commented on tests/kafkatest/tests/streams/streams_static_membership_test.py:
```diff
@@ -50,8 +55,10 @@ def __init__(self, test_context):
 
     @cluster(num_nodes=8)
-    def test_rolling_bounces_will_not_trigger_rebalance_under_static_membership(self):
-        self.zookeeper.start()
```
Should we do this change in `streams_cooperative_rebalance_upgrade_test` also?

AlanConfluent replied:
I'm doing that in a followup PR

guozhangwang reviewed (commented):
Thanks @AlanConfluent ! A few meta questions:

1) Seems the `streams_cooperative_rebalance_upgrade_test.py` is not included in this PR?
2) I think for `streams_application_upgrade_test.py` we should also consider enabling kraft on the servers, to make sure that kraft works when streams itself upgrade.
3) In `streams_broker_compatibility_test.py` when we test for broker versions > 3.1 we should also allow it to be kraft.
4) This is not related to this PR, but it seems the test coverage for `streams_application_ugprade_test` and `streams_upgrade_test` have much overlaps. @vvcephei could you chime in here since you have much experience with the former class file. Could we dedup their coverage hence reduce our test time?
5) If the current e2e time is too high, I feel maybe we can skip adding the kraft model for the following suite:
a. shutdown_deadlock
b. relational_smoke, since its dependency on kraft is exactly covered by smoke -- i.e. if there's an issue with relational_smoke due to kraft, then smoke itself should fail as well.
c. named_repartition

AlanConfluent commented:
> > In addition to addressing the review comments, can you post a link to the system test results with this change? It would be good to verify the impact before merging.
> 
> +1. I'd also love to learn how much system test time increase this one would incur.

My main link is here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/5067/

I'm trying to run the same tests on trunk and see, but there are unrelated failures.  Will update when I get a run.

AlanConfluent commented:
> Thanks @AlanConfluent ! A few meta questions:
> 
> 1. Seems the `streams_cooperative_rebalance_upgrade_test.py` is not included in this PR?

I didn't update any of the upgrade tests just yet.  I was having a hard time running against different versions of Kafka and using KRaft -- this seems like a reasonable followup, so I didn't do it yet.

> 2. I think for `streams_application_upgrade_test.py` we should also consider enabling kraft on the servers, to make sure that kraft works when streams itself upgrade.

I agree.  This will be a followup.

> 3. In `streams_broker_compatibility_test.py` when we test for broker versions > 3.1 we should also allow it to be kraft.

Maybe this was the main issue I was running into.  I thought KRaft was available in earlier versions, but saw odd failures.  I'll talk to you about what it would take to get this working.


> 4. This is not related to this PR, but it seems the test coverage for `streams_application_ugprade_test` and `streams_upgrade_test` have much overlaps. @vvcephei could you chime in here since you have much experience with the former class file. Could we dedup their coverage hence reduce our test time?

> 5. If the current e2e time is too high, I feel maybe we can skip adding the kraft model for the following suite:
>    a. shutdown_deadlock
>    b. relational_smoke, since its dependency on kraft is exactly covered by smoke -- i.e. if there's an issue with relational_smoke due to kraft, then smoke itself should fail as well.
>    c. named_repartition
Good to consider.  Let's evaluate after I get a successful run of trunk.

AlanConfluent committed:
Mention zk as a matrix param when run using default for bounce test

guozhangwang commented:
Re-triggered the jenkins build.

AlanConfluent committed:
Run only quorum.remote_kraft for most tests

AlanConfluent commented:
I made the change to switch some of these tests to just run with remote_kraft to minimize test run time.

vvcephei reviewed (approved):
Thanks, @AlanConfluent !

vvcephei commented:
It looks like you have a green run: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/system-test-kafka-branch-builder--1661474350--AlanConfluent--updates_tests_kraft--9be74f3d1/2022-08-25--001./2022-08-25--001./report.html

AlanConfluent commented:
I think this was the ultimate run which covered everything under streams:
http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/system-test-kafka-branch-builder--1661556003--AlanConfluent--updates_tests_kraft--9be74f3d1/2022-08-26--001./2022-08-26--001./report.html
